manager/volume.go

```diff
@@ -765,21 +775,86 @@ func (m *VolumeManager) CancelExpansion(volumeName string) (v *longhorn.Volume,
 	return v, nil
 }
 
-func (m *VolumeManager) UpdateRecurringJobs(volumeName string, jobs []types.RecurringJob) (v *longhorn.Volume, err error) {
+func (m *VolumeManager) AddVolumeRecurringJob(volumeName string, name string, isGroup bool) (volumeRecurringJob map[string]*types.VolumeRecurringJob, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "unable to update volume recurring jobs for %v", volumeName)
+	}()
+
+	v, err := m.ds.GetVolume(volumeName)
+	if err != nil {
+		return nil, err
+	}
+
+	key := ""
+	if isGroup {
+		key = types.GetRecurringJobLabelKey(types.LonghornLabelCronJobGroup, name)
+	} else {
+		key = types.GetRecurringJobLabelKey(types.LonghornLabelCronJob, name)
+	}
+	v.Labels[key] = types.LonghornLabelValueEnabled
+	v, err = m.ds.UpdateVolume(v)
+	if err != nil {
+		return nil, err
+	}
+	volumeRecurringJob, err = m.ListVolumeRecurringJob(volumeName)
+	if err != nil {
+		return nil, err
+	}
+	logrus.Debugf("Updated volume %v recurring jobs to %+v", v.Name, volumeRecurringJob)
+	return volumeRecurringJob, nil
+}
+
+func (m *VolumeManager) ListVolumeRecurringJob(volumeName string) (map[string]*types.VolumeRecurringJob, error) {
+	var err error
 	defer func() {
 		err = errors.Wrapf(err, "unable to update volume recurring jobs for %v", volumeName)
 	}()
 
-	if err = m.validateRecurringJobs(jobs); err != nil {
+	v, err := m.ds.GetVolume(volumeName)
+	if err != nil {
 		return nil, err
 	}
 
+	groupPrefix := fmt.Sprintf(types.LonghornLabelRecurringJobKeyPrefixFmt, types.LonghornLabelCronJobGroup) + "/"
+	jobPrefix := fmt.Sprintf(types.LonghornLabelRecurringJobKeyPrefixFmt, types.LonghornLabelCronJob) + "/"
+	jobMapVolumeJob := make(map[string]*types.VolumeRecurringJob)
+	for label := range v.Labels {
+		if strings.HasPrefix(label, groupPrefix) {
+			jobName := strings.TrimPrefix(label, groupPrefix)
+			jobMapVolumeJob[jobName] = &types.VolumeRecurringJob{
+				Name:    jobName,
+				IsGroup: true,
+			}
+			continue
+		} else if strings.HasPrefix(label, jobPrefix) {
+			jobName := strings.TrimPrefix(label, jobPrefix)
+			jobMapVolumeJob[jobName] = &types.VolumeRecurringJob{
+				Name:    jobName,
+				IsGroup: false,
+			}
+		}
+	}
+
+	return jobMapVolumeJob, nil
+}
+
+func (m *VolumeManager) DeleteVolumeRecurringJob(volumeName string, name string, isGroup bool) (v *longhorn.Volume, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "unable to update volume recurring jobs for %v", volumeName)
+	}()
+
 	v, err = m.ds.GetVolume(volumeName)
 	if err != nil {
 		return nil, err
 	}
 
-	v.Spec.RecurringJobs = jobs
+	key := ""
+	if isGroup {
+		key = types.GetRecurringJobLabelKey(types.LonghornLabelCronJobGroup, name)
+	} else {
+		key = types.GetRecurringJobLabelKey(types.LonghornLabelCronJob, name)
+	}
+	delete(v.Labels, key)
```

Comment:
Check if it exists before deletion.

Reply:
Is this necessary?
[The delete built-in function deletes the element with the specified key (m[key]) from the map. If m is nil or there is no such element, delete is a no-op.](https://pkg.go.dev/builtin#delete)

Reply:
Thanks for @innobead explanation. The comment is referring to the following `UpdateVolume`.

Reply:
Fixed.